### PR TITLE
[SPARK-38047][K8S] Add `OUTLIER_NO_FALLBACK` executor roll policy

### PR DESCRIPTION
**Pull request by alexholmes**
### What changes were proposed in this pull request?

This PR aims to add a new executor roll policy which allows users to skip rolling in cases where there are no outlier executors.

### Why are the changes needed?

As currently implemented an executor is always rolled every `spark.kubernetes.executor.rollInterval` interval. In environments where starting of executors can introduce latencies it may be desirable for users to have the option to determine if rolling should only happen when outliers are found.

### Does this PR introduce _any_ user-facing change?

No, this is an additional option being added to a new feature in Apache Spark 3.3.

### How was this patch tested?

Pass the CIs with the newly added test cases.

### CODE REVIEW

**Review by dongjoon-hyun — COMMENTED** [review]
Welcome to the Apache Spark community, @alexholmes .

**Comment by dongjoon-hyun** [comment]
Let me close and reopen this to trigger your GitHub Action properly.

**Comment by alexholmes** [comment]
Thank you @dongjoon-hyun !

**Comment by dongjoon-hyun** [comment]
It seems that your GitHub Action is not responding still for some reason.
- https://github.com/alexholmes/spark/actions/workflows/build_and_test.yml

**Comment by dongjoon-hyun** [comment]
Could you do force-push this on your branch?

**Comment by dongjoon-hyun** [comment]
FYI, I revised your PR title according to the community style, `[SPARK-38047][K8S]`. Each Apache community has different style unfortunately.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
BTW, I'd like to recommend not to touch the other policy, @alexholmes .
They are orthogonal to the outlier. Please make a new policy for your requirement.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala [hunk 1]
```diff
@@ -182,6 +182,16 @@ private[spark] object Config extends Logging {
       .checkValue(_ >= 0, "The minimum number of tasks should be non-negative.")
       .createWithDefault(0)
 
+  val EXECUTOR_ROLL_OUTLIERS_ONLY =
+    ConfigBuilder("spark.kubernetes.executor.onlyRollOutliers")
+      .doc("Only roll an executor if it is an outlier. An outlier is defined to be at least two " +
+        "standard deviation outside of the mean. If no outlier is found then no executor will " +
+        "be rolled."
+      )
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:190]: Please place `)` at line 189.

Comment (alexholmes) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala]: I removed the config so I'm going to resolve this.

**Comment by alexholmes** [comment]
@dongjoon-hyun I've updated the PR with the following changes:

1. Introduced a new policy, OUTLIER_OR_TOTAL_DURATION, which provides a fallback of TOTAL_DURATION policy for outliers (basically the prior implementation of OUTLIER).
2. Updated the OUTLIER policy to not provide a fallback, therefore allowing no executors to be rolled when there are no outliners

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala [hunk 2]
```diff
@@ -165,7 +166,9 @@ private[spark] object Config extends Logging {
         "OUTLIER policy chooses an executor with outstanding statistics which is bigger than" +
         "at least two standard deviation from the mean in average task time, " +
         "total task time, total task GC time, and the number of failed tasks if exists. " +
-        "If there is no outlier, it works like TOTAL_DURATION policy.")
+        "If there is no outlier then no executor will be rolled." +
+        "OUTLIER_OR_TOTAL_DURATION policy picks an outlier using the OUTLIER policy above. " +
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:170]: Could you keep the previous behavior by making `OUTLIER_OR_TOTAL_DURATION` as the default?
```scala
- .createWithDefault(ExecutorRollPolicy.OUTLIER.toString)
+ .createWithDefault(ExecutorRollPolicy.OUTLIER_OR_TOTAL_DURATION.toString) 
```

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPlugin.scala [hunk 3]
```diff
@@ -117,29 +117,36 @@ class ExecutorRollDriverPlugin extends DriverPlugin with Logging {
...
-        // outliers in various perspective:
-        //   AVERAGE_DURATION > TOTAL_DURATION > TOTAL_GC_TIME > FAILED_TASKS
-        // Since we will choose only first item, the duplication is okay. If there is no outlier,
-        // We fallback to TOTAL_DURATION policy.
-        outliers(listWithoutDriver.filter(_.totalTasks > 0), e => e.totalDuration / e.totalTasks) ++
-          outliers(listWithoutDriver, e => e.totalDuration) ++
-          outliers(listWithoutDriver, e => e.totalGCTime) ++
-          outliers(listWithoutDriver, e => e.failedTasks) ++
-          listWithoutDriver.sortBy(_.totalDuration).reverse
+      case ExecutorRollPolicy.OUTLIER => outliers(listWithoutDriver)
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPlugin.scala:120]: If you don't mind, please put `outliers(listWithoutDriver)` at the next line.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPlugin.scala [hunk 4]
```diff
@@ -117,29 +117,36 @@ class ExecutorRollDriverPlugin extends DriverPlugin with Logging {
...
+        // If there is no outlier we fallback to TOTAL_DURATION policy.
+        outliers(listWithoutDriver) ++ listWithoutDriver.sortBy(_.totalDuration).reverse
     }
     sortedList.headOption.map(_.id)
   }
 
+  /**
+   * We build multiple outlier lists and concat in the following importance order to find
+   * outliers in various perspective:
+   * AVERAGE_DURATION > TOTAL_DURATION > TOTAL_GC_TIME > FAILED_TASKS
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPlugin.scala:131]: Could you add two more leading space here?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPlugin.scala [hunk 5]
```diff
@@ -117,29 +117,36 @@ class ExecutorRollDriverPlugin extends DriverPlugin with Logging {
...
+   * Since we will choose only first item, the duplication is okay.
+   */
+  private def outliers(listWithoutDriver: Seq[v1.ExecutorSummary]) =
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPlugin.scala:134]: `outliers` -> `outliersFromMultipleDimentions`?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPlugin.scala [hunk 6]
```diff
@@ -117,29 +117,36 @@ class ExecutorRollDriverPlugin extends DriverPlugin with Logging {
...
+
   /**
    * Return executors whose metrics is outstanding, '(value - mean) > 2-sigma'. This is
    * a best-effort approach because the snapshot of ExecutorSummary is not a normal distribution.
    * Outliers can be defined in several ways (https://en.wikipedia.org/wiki/Outlier).
    * Here, we borrowed 2-sigma idea from https://en.wikipedia.org/wiki/68-95-99.7_rule.
    * In case of normal distribution, this is known to be 2.5 percent roughly.
    */
-  private def outliers(
+  private def outliersForFn(
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPlugin.scala:149]: Let's keep the original method name because this is focusing on a single dimension outlier.

**Review by dongjoon-hyun — COMMENTED** [review]
Could you revise the PR title and description according to the new PR content?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala [hunk 7]
```diff
@@ -147,7 +147,8 @@ private[spark] object Config extends Logging {
       .createWithDefault(0)
 
   object ExecutorRollPolicy extends Enumeration {
-    val ID, ADD_TIME, TOTAL_GC_TIME, TOTAL_DURATION, AVERAGE_DURATION, FAILED_TASKS, OUTLIER = Value
+    val ID, ADD_TIME, TOTAL_GC_TIME, TOTAL_DURATION, AVERAGE_DURATION, FAILED_TASKS,
+    OUTLIER, OUTLIER_OR_TOTAL_DURATION = Value
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:151]: indentation?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala [hunk 8]
```diff
@@ -165,7 +166,9 @@ private[spark] object Config extends Logging {
         "OUTLIER policy chooses an executor with outstanding statistics which is bigger than" +
         "at least two standard deviation from the mean in average task time, " +
         "total task time, total task GC time, and the number of failed tasks if exists. " +
-        "If there is no outlier, it works like TOTAL_DURATION policy.")
+        "If there is no outlier then no executor will be rolled." +
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:169]: We need a space at the end of the string. `rolled. "`

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPlugin.scala [hunk 9]
```diff
@@ -117,29 +117,36 @@ class ExecutorRollDriverPlugin extends DriverPlugin with Logging {
...
+  /**
+   * We build multiple outlier lists and concat in the following importance order to find
+   * outliers in various perspective:
+   * AVERAGE_DURATION > TOTAL_DURATION > TOTAL_GC_TIME > FAILED_TASKS
+   * Since we will choose only first item, the duplication is okay.
+   */
+  private def outliers(listWithoutDriver: Seq[v1.ExecutorSummary]) =
+    outliersForFn(
+      listWithoutDriver.filter(_.totalTasks > 0),
+      e => e.totalDuration / e.totalTasks) ++
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPlugin.scala:137]: Please merge line 135 ~ 137 as a one liner if possible.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPluginSuite.scala [hunk 10]
```diff
@@ -132,7 +132,9 @@ class ExecutorRollPluginSuite extends SparkFunSuite with PrivateMethodTester {
   }
 
   test("A one-item executor list") {
-    ExecutorRollPolicy.values.foreach { value =>
+    ExecutorRollPolicy.values
+      .filter(_ != ExecutorRollPolicy.OUTLIER)
+      .foreach { value =>
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPluginSuite.scala:137]: The following would be shorter and reasonable.
```scala
ExecutorRollPolicy.values.filter(_ != ExecutorRollPolicy.OUTLIER).foreach { value =>
```

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPluginSuite.scala [hunk 11]
```diff
@@ -172,10 +174,53 @@ class ExecutorRollPluginSuite extends SparkFunSuite with PrivateMethodTester {
...
 
-  test("Policy: OUTLIER - Work like TOTAL_DURATION if there is no outlier") {
+  test("Policy: OUTLIER_OR_TOTAL_DURATION - Use TOTAL_DURATION if there is no outlier") {
     assertEquals(
       plugin.invokePrivate(_choose(list, ExecutorRollPolicy.TOTAL_DURATION)),
-      plugin.invokePrivate(_choose(list, ExecutorRollPolicy.OUTLIER)))
+      plugin.invokePrivate(_choose(list, ExecutorRollPolicy.OUTLIER_OR_TOTAL_DURATION)))
+  }
+
+  test("Policy: OUTLIER_OR_TOTAL_DURATION - Detect an average task duration outlier") {
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPluginSuite.scala:183]: Actually, this is orthogonal to your proposal. We usually put one theme in a PR and avoid mixing different things.
Could you make a separate PR for adding more test coverage for `OUTLIER_OR_TOTAL_DURATION`?

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPluginSuite.scala [hunk 12]
```diff
@@ -172,10 +174,53 @@ class ExecutorRollPluginSuite extends SparkFunSuite with PrivateMethodTester {
...
+      1000, 0, 0,
+      0, false, 0, new Date(1639300001000L),
+      Option.empty, Option.empty, Map(), Option.empty, Set(), Option.empty, Map(), Map(), 1,
+      false, Set())
+    assertEquals(
+      plugin.invokePrivate(_choose(list :+ outlier, ExecutorRollPolicy.TOTAL_GC_TIME)),
+      plugin.invokePrivate(_choose(list :+ outlier, ExecutorRollPolicy.OUTLIER_OR_TOTAL_DURATION)))
+  }
+
+  test("Policy: OUTLIER - Return None if not outliers") {
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPluginSuite.scala:222]: `not outliers` -> `there is no outlier`.

Comment (dongjoon-hyun) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPluginSuite.scala:222]: Please move this test case after `test("A one-item executor list")`.

**Comment by alexholmes** [comment]
@dongjoon-hyun Thanks for your feedback. I've made the following changes:

1. Kept the behavior of OUTLIERS policy unchanged, along with its associated unit tests.
2. Introduced a new OUTLIER_NO_FALLBACK policy which doesn't fallback to TOTAL_DURATION in cases where there are no outliers
3. Updated the title and content of this PR to reflect these changes
4. Addressed your formatting comments above

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you for updating, @alexholmes .
Merged to master for Apache Spark 3.3.0.
